components/brave_rewards/browser/diagnostic_log.cc

```diff
@@ -0,0 +1,353 @@
+  }
+
+  const int64_t size = file.GetLength() - offset;
+  std::unique_ptr<char[]> buffer = std::make_unique<char[]>(size + 1);
+
+  if (file.ReadAtCurrentPos(buffer.get(), size) == -1) {
+    return "";
+  }
+
+  return std::string(buffer.get());
+}
+
+bool WriteOnFileTaskRunner(const base::FilePath& file_path,
+                           const std::string& log_entry,
+                           int64_t max_file_size,
+                           int keep_num_lines,
+                           bool first_write) {
+  base::File file;
+  if (!CreateOrOpen(file_path, &file)) {
+    return false;
+  }
+
+  if (file.Seek(base::File::FROM_END, 0) == -1) {
+    return false;
+  }
+
+  if (first_write) {
+    const std::string divider = std::string(kDividerLength, '-') + "\n";
+    file.WriteAtCurrentPos(divider.c_str(), divider.length());
+  }
+
+  if (file.WriteAtCurrentPos(log_entry.c_str(), log_entry.length()) == -1) {
+    return false;
+  }
+
+  return MaybeTrimBeginningOfFile(&file, max_file_size, keep_num_lines,
+                                  first_write);
+}
+
+}  // namespace
+
+namespace brave_rewards {
+
+DiagnosticLog::DiagnosticLog(const base::FilePath& file_path,
+                             int64_t max_file_size,
+                             int keep_num_lines)
+    : file_task_runner_(base::CreateSequencedTaskRunner(
+          {base::ThreadPool(), base::MayBlock(),
+           base::TaskPriority::USER_VISIBLE,
+           base::TaskShutdownBehavior::BLOCK_SHUTDOWN})),
+      file_path_(file_path),
+      max_file_size_(max_file_size),
+      keep_num_lines_(keep_num_lines),
+      first_write_(true) {}
+
+DiagnosticLog::~DiagnosticLog() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+}
+
+void DiagnosticLog::ReadLastNLines(int num_lines, ReadCallback callback) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  file_task_runner_->PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(&ReadLastNLinesOnFileTaskRunner, file_path_, num_lines),
+      base::BindOnce(&DiagnosticLog::OnReadLastNLines, AsWeakPtr(),
+                     std::move(callback)));
+}
+
+void DiagnosticLog::Write(const std::string& log_entry,
+                          WriteCallback callback) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  file_task_runner_->PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(&WriteOnFileTaskRunner, file_path_, log_entry,
+                     max_file_size_, keep_num_lines_, first_write_),
+      base::BindOnce(&DiagnosticLog::OnWrite, AsWeakPtr(),
+                     std::move(callback)));
+  first_write_ = false;
+}
+
+void DiagnosticLog::Write(const std::string& log_entry,
+                          const base::Time& time,
+                          const std::string& file,
+                          int line,
+                          int verbose_level,
+                          WriteCallback callback) {
+  const std::string formatted_time = FormatTime(time);
```

Comment:
shouldn't this method also have a sequence check?

Reply:
Yes, fixed.